Transpose pdll

KavithaTipturMadhu opened the pull request:
tests pending:pr1
Edit: I have updated the PR with unit tests and rebased it on top of the new pipeline, its good for review.
Edit2: Removed the transpose and gemm code, need to rebase on top of the bundle pr, will convert back to non-draft mode once that is done and tests are added.
Edit3: moved to new pipeline,  unit tests and integration tests updated.

rengolin reviewed (commented):
Some initial comments

rengolin commented on lib/TPP/DefaultTppPasses.cpp:
```diff
@@ -115,11 +115,13 @@ struct DefaultTppPasses
-
-        // TODO: Add a flag for this pass so it doesn't conflate with other options.
-        pm.addNestedPass<func::FuncOp>(createVectorContractToOuterproduct());
+        if (outerProduct) {
```
After #983 you can just add your stuff to the `VectorToXSMM` bundle.

rengolin commented on include/TPP/Conversion/ConvertVectorToXsmm/ConvertTranspose.h:
```diff
@@ -0,0 +1,9 @@
+
+namespace mlir {
+namespace tpp {
+std::unique_ptr<mlir::Pass> createInsertTranspose();
```
Actually, scratch that. This should reuse the bundle introduced in #983.

alheinecke commented on include/TPP/Transforms/Utils/VNNIUtils.h:
```diff
@@ -40,12 +45,19 @@ std::optional<int64_t> getVnniBlockingFactor(Type type);
 bool isInVnniLayout(VnniOperandRank expectedRank, MemRefType memref);
 
+// Return true if the memref is in VNNI layout with rank `expectedRank`.
+bool isInVnniLayout(VnniOperandRank expectedRank, VectorType vector);
```
The we need VNNI Utils for transpose. I would suggest adding Norm->VNNI function call lowering in a different PR to keep PRs small.

KavithaTipturMadhu replied:
We need this for when contraction inputs are not compatible to be lowered into brgemm. Should we seperate that out into a different patch? If yes, I'll move this out and keep just the transpose lowering in this patch for now. That makes the pr quite clean and small.

rengolin replied:
> If yes, I'll move this out and keep just the transpose lowering in this patch for now.

Yes, please. Keep here _only_ transpose logic.

alheinecke commented on include/TPP/Dialect/Xsmm/XsmmUtils.h:
```diff
@@ -21,6 +26,22 @@ class ArrayAttr;
 
 namespace xsmm {
+
+struct BrgemmInfo {
```
Why do we need the BRGEMM Info struct when lowering Transpose.

KavithaTipturMadhu replied:
Agreed, will remove this. It got left out while moving the code.

alheinecke commented on include/TPP/Dialect/Xsmm/XsmmUtils.h:
```diff
@@ -77,6 +104,9 @@ FailureOr<UnaryFlags> getUnaryFlags(Type inputType, Type outputType);
 
 // Compute the broadcasting flags for 'operandType' based on 'outputType'.
 enum class OperandPos { LHS = 0, RHS = 1 };
+FailureOr<BinaryFlags> getBinFlags(ArrayRef<int64_t> shapeOutput,
```
Transpose is unary, no addition on Binary is needed.

KavithaTipturMadhu replied:
Same comment as above, will remove.

alheinecke commented on include/TPP/Dialect/Xsmm/XsmmUtils.h:
```diff
@@ -92,6 +122,40 @@ template <typename DispatchOpTy>
+std::optional<unsigned>
+getPosInCodomain(unsigned dim, vector::ContractionOp contractOp, AffineMap map);
+
+LogicalResult checkAccess(PatternRewriter &rewriter,
```
This seems to be GEMM related as well, no need in my eyes.

KavithaTipturMadhu replied:
Will remove, same comment as above.

alheinecke commented on include/TPP/Dialect/Xsmm/XsmmUtils.h:
```diff
@@ -92,6 +122,40 @@ template <typename DispatchOpTy>
+                             unsigned n, unsigned k, xsmm::DataTypeAttr type);
+
+std::optional<unsigned>
+getPosInCodomain(unsigned dim, vector::ContractionOp contractOp, AffineMap map);
```
This seems to be GEMM related as well, no need in my eyes.

KavithaTipturMadhu replied:
same comment as above.

alheinecke commented on lib/TPP/Conversion/ConvertVectorToXsmm/ConvertTranspose.cpp:
```diff
@@ -0,0 +1,405 @@
+    PatternRewriter rewriter(&getContext());
+    // Enable conversion for linalg.generic to XSMM Brgemm if possible.
+    auto res =
+        getOperation()->walk([&](mlir::vector::ContractionOp contractOp) {
```
We want to rewrite transpose, why are we walking over contraction ops?

KavithaTipturMadhu replied:
Like i said earlier, I thought I'd have the code to insert transposes when gemms cannot be performed in this patch itself. If we want to keep this in the brgemm/gemm lowering patch, I'll move this to the next patch.

alheinecke reviewed (commented):
It seems there is also some BRGEMM matching in this PR which I don't understand. I thought the idea is to rewrite transpose.

rengolin commented on lib/TPP/GPU/GpuVectorize.cpp:
```diff
@@ -29,6 +29,7 @@
 #include "mlir/Transforms/Passes.h"
 
 using namespace mlir;
+using namespace mlir::vector;
```
Why is this here? Please keep the unrelated changes for further PRs. This carries the `mlir::vector` changes everywhere that are not necessary.

KavithaTipturMadhu replied:
Its not an unrelated change, it is a build fix, vector is not disambiguated otherwise. std::vector and mlir:vector are both visible to the .cpp file. We need to explicitly make mlir::vector visible for build to go through.

rengolin replied:
This should only happen if both `std` and `mlir` namespaces are in scope. You should not have `using namespace std` in LLVM source files.

rengolin commented on lib/TPP/Conversion/ConvertVectorToXsmm/ConvertTranspose.cpp:
```diff
@@ -0,0 +1,299 @@
+        rewriter, output, module, inputs, preceedingOperands, -1, operandRange,
+        invokeName, dtype, true);
+
+    return std::make_pair(&*dispatched, &*invokeCall);
```
This is a repeat of the code below. Please condense the two transposes into one function with different arguments for VNNI and non-VNNI.

KavithaTipturMadhu commented on include/TPP/Passes.h:
```diff
@@ -98,6 +98,8 @@ class XeGPUDialect;
 
 } // namespace mlir
 
+#include "TPP/Conversion/ConvertVectorToXsmm/ConvertTranspose.h"
```
We need to make the headers visible to tpp-opt somehow in order to make the passes available in standalone mode. We can include them directly in both tpp-opt and in VectorToXsmm.cpp, or we could include them in Passes.h. I thought it made sense to include them in Passes.h because  they are passes after all, and they ought to be visible wherever the rest of the passes are, but if you think that needn't be, we can do what I said prreviously i.e., include the headers in two places.

KavithaTipturMadhu commented on include/TPP/Passes.td:
```diff
@@ -54,7 +54,18 @@ def VectorizationPass : Pass<"vectorization-pass",
+  let dependentDialects = [ "memref::MemRefDialect", "linalg::LinalgDialect", "mlir::vector::VectorDialect" ];
+}
+
+def InsertTranspose : Pass<"insert-transpose-pass", "func::FuncOp"> {
```
This pass just converts a transpose, the next pass converts the rest of the constructs to xsmm. Isn't it an overkill to match transpose pattern in the pass and then invoke the pass, when the pass itself does that for you? The pass is currently translating just transposes, nothing else, so it pattern matches only  transposes. It achieves the same effect as having to match transpose pattern and lowering only those.

rengolin replied:
We're talking about public names. The final objective of the pass is to convert the whole set of operations, even if it just does transposes for now. We should not change the name of the pass, classes and functions as we add more functionality.

KavithaTipturMadhu replied:
As discusssed, I'm not implementing transposes for brgemms for now. Hence, merging transposes and other vector ops in one and renaming these pdll and cpp files as ConvertVectorToXsmm.

KavithaTipturMadhu commented on lib/TPP/DefaultPipeline.cpp:
```diff
@@ -64,17 +64,19 @@ llvm::cl::opt<bool> lowerPackUnpackWithoutTranspose(
+            llvm::cl::list_init<unsigned>(SmallVector<unsigned>{8, 16}),
+            llvm::cl::CommaSeparated);
+
+llvm::cl::opt<bool> vectorToXSMM("vector-To-XSMM",
```
I tried to use the same flag as the one in passbundles, I get an error saying there's no such option in tpp-run. However, if I try to add another command with the same name, it complains that the command already exists. I didn't know to resolve it immediately and added a command with a different name for now.

rengolin replied:
Please don't just create a new option if you can't use an existing one.

For `tpp-opt`, because you have to pass `-default-tpp-passes`, you need to add the flag to that one, for example `-default-tpp-passes="vector-to-xsmm"`. In `tpp-run`, because the default TPP passes will run by default, you can just call `-vector-to-xsmm`.

KavithaTipturMadhu replied:
I tried invoking tpp-run with --vector-to-xsmm. I get an error "Unknown command line argument". Can you please let me know how to pass vector-to-xsmm flag to tpp-run? I don't see it in help either.

KavithaTipturMadhu replied:
Can you please tell me how to pass the right argument? I cant run tests and I have no clue how to pass the argument, I tried everything, all variants of vector-to-xsmm, nothing seems to work.

KavithaTipturMadhu replied:
I had to add the argument to tpp-run.cpp, just like gpuBackend commandline argument, and that is working.

KavithaTipturMadhu commented on lib/TPP/Conversion/ConvertVectorToXsmm/ConvertTranspose.cpp:
```diff
@@ -0,0 +1,299 @@
+      return failure(transposeOp);
+    }
+  }
+  return success(transposeOp);
```
I cant return just transposeOp because that invokes the private constructor of LogicalResult and results in a compiler error.

rengolin replied:
I'm not sure what you expect passing the op will do. I expected you wanted the actual operations, but if you really just want the `LogicalResult`, then adding arguments will convert them to a `bool` and change the meaning of `success` / `failure` if the argument converts to `true` or `false`.

KavithaTipturMadhu replied:
Change the meaning how? The diagnostic engine is expected to propagate the value of transposeop, not change the return value of success/failure, yes? Unless my understanding is completely wrong.

KavithaTipturMadhu commented on lib/TPP/Conversion/ConvertVectorToXsmm/ConvertTranspose.cpp:
```diff
@@ -0,0 +1,299 @@
+            dyn_cast<mlir::vector::TransposeOp>(transposeOp))) {
+      return failure(transposeOp);
+    }
+    auto unaryInfo = xsmm::utils::getVectorUnaryInfo(
```
This was actually a mistake on my part, we dont need to check unary info call in case of vnni transpose, I have fixed the validation call.

rengolin replied:
Code is still duplicated, at least on PR, please don't mark as done until we can all see the changes.

KavithaTipturMadhu replied:
Which code are you referring to? I thought you meant the unary info call.

KavithaTipturMadhu replied:
@rengolin can you please help me understand what I am missing here?

KavithaTipturMadhu commented on lib/TPP/Conversion/ConvertVectorToXsmm/ConvertTranspose.cpp:
```diff
@@ -0,0 +1,299 @@
+namespace mlir {
+namespace tpp {
+
+struct ConvertTranspose
```
Actually, this one is just the transpose conversion, because we want to prioritize transpose conversions over other ops, and we don't know how else to force conversion of transposes first before converting other ops. That's the reason transposes are in a separate pdll and the rest of the ops are in a subsequent pdll.

rengolin replied:
> we want to prioritize transpose conversions over other ops

I'm not sure about this. Can you expand?

> That's the reason transposes are in a separate pdll

I understood we were just adding transpose first because it's easier and allows us to review the infrastructure before the complex ops. I don't see any special reason why the transpose should be before any other.

But even if so, there has to be an order to PDL patterns that is not individually running them. For instance, we want to look at fused GEMMs before plain GEMMs, softmax before a sequence of element-wise and reduction, etc.

KavithaTipturMadhu replied:
As discussed, we'll be moving transposes and other vector ops into a single pdll file for now.

KavithaTipturMadhu commented on lib/TPP/Conversion/ConvertVectorToXsmm/ConvertTransposePDL.pdll:
```diff
@@ -0,0 +1,22 @@
+#include "mlir/IR/BuiltinTypes.td"
+
+
+Rewrite BuildTranspose(op:Op<vector.transpose>, input0:Op<vector.transfer_read>, output: Op<vector.transfer_write>, outputType:TypeRange)->(dispatch:Op<func.callOp>, invoke:Op<func.callOp>);
```
As I mentioned, this one is just transpose converter and not general xsmm cnverter, the brgemm patch has the general xsmm converter :  https://github.com/plaidml/tpp-mlir/pull/985

KavithaTipturMadhu commented on lib/TPP/Conversion/ConvertVectorToXsmm/ConvertTranspose.cpp:
```diff
@@ -0,0 +1,299 @@
+  operandRange.push_back(input->getOperand(0));
+  operandRange.push_back(output->getOperand(1));
+  SmallVector<Value> inputs;
+  SmallVector<Value> preceedingOperands;
```
The function takes smallvector arguments, however, I can initialize directly without push backs, I have changed that.

rengolin replied:
I can't see the changes here, did you push them to the branch?

KavithaTipturMadhu replied:
I thought you were referring to the operandRange SmallVector. I will update the other arguments.

KavithaTipturMadhu replied:
I have updated the function to take just one SmallVector of arguments, I have kept it as SmallVector initialized statically and hence it gets allocated on stack.

rengolin commented on include/TPP/Dialect/Xsmm/XsmmUtils.h:
```diff
@@ -88,6 +98,28 @@ ArrayAttr getUnaryDispatchFlags(UnaryOp op);
 
 ArrayAttr getBinaryDispatchFlags(BinaryOp op);
 
+int64_t getOredFlags(ArrayAttr flags);
```
This already exists in `ConvertXsmmToFunc`, please reuse there.

KavithaTipturMadhu replied:
I did not want to add dependences between the two conversion modules, and we were planning to retire ConvertXsmmToFunc after this. But if you suggest that I reuse this function, I will for now, We can move it to xsmmutils when we retire it, I suppose.

KavithaTipturMadhu replied:
Not depending on ConvertXsmmToFunc now, because this function is no longer required.

rengolin commented on lib/TPP/Conversion/ConvertVectorToXsmm/CMakeLists.txt:
```diff
@@ -0,0 +1,44 @@
+add_mlir_pdll_library (
```
Do we really need one PDLL library per pattern? This sounds like a huge scalability problem.

KavithaTipturMadhu replied:
I didn't say we need one per pattern, we just need one for transpose and then one for the rest of them to make sure transpose runs before the others do. This is so that we can insert and translate transposes before gemms/brgemms.

KavithaTipturMadhu replied:
As discussed, we'll be moving transposes and other vector ops into a single pdll file for now.

rengolin commented on lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:
```diff
@@ -8,18 +8,67 @@
+#include <map>
+#define DEBUG_TYPE "xsmm-utils"
+
+using namespace mlir;
```
this is unnecessary, as below you're already in the `mlir` namespace.

Please don't use `using namespace` in these files, as it complicates namespace resolution. Use the existing pattern of entering and leaving namespaces as needed.

KavithaTipturMadhu replied:
Removed the namespace mlir in XsmmUtils.cpp. However, I haven't removed it in other files because that meant namespace resolution for a bunch of objects there.

rengolin commented on lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:
```diff
@@ -287,6 +358,175 @@ FailureOr<BinaryFlags> getBinaryFlags(Type operandType, Type outputType,
+  return strides[pos];
+}
+
+bool isTwoDTransposeOp(vector::TransposeOp transposeOp) {
```
Sounds like a simpler approach would be to create a function to get the rank of a vector (in view of VNNI layout, which just needs the operand).

KavithaTipturMadhu replied:
This funciton belongs in the brgemm patch. Removing this.

rengolin replied:
you still seem to be using in the code? Is this really just for the brgemm patch?

KavithaTipturMadhu replied:
Oh i thought you were referring to the line above it, you were referring to the twoDtransposeOp. I will update it to use ranks.

rengolin commented on lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:
```diff
@@ -287,6 +358,175 @@ FailureOr<BinaryFlags> getBinaryFlags(Type operandType, Type outputType,
+  bool isVnni = vnni::utils::isInVnniLayout(operandType.getRank(), operandType);
+  if (isVnni || operandType.getRank() != 2 ||
+      dyn_cast<VectorType>(transposeOp.getResult().getType()).getRank() != 2 ||
+      (isa<scf::ForallOp>(transposeOp->getParentOp()) &&
```
This is unstable, as the parent op could be `scf::parallel` or `scfp:for` or `func::func` and you'd have to multiply the checks here.

KavithaTipturMadhu replied:
We run scf:foralltoparallelloop conversion after this pass. As for the funcop, its taken care of in this check already.

rengolin commented on lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:
```diff
@@ -287,6 +358,175 @@ FailureOr<BinaryFlags> getBinaryFlags(Type operandType, Type outputType,
+  return results;
+}
+
+int64_t getOredFlags(ArrayAttr flags) {
```
This sounds like a GEMM function, but you use for the unary check below. I'm not sure what's going on.

KavithaTipturMadhu replied:
I realize that, the flag is always unary::none so we don't need oredflags function. I'll get rid of it.

KavithaTipturMadhu replied:
Just as a side note, getOredFlags is used on unaryFlags in ConvertXsmmToFunc which is why I ended up using the same utility. We'll need to update that as well, till such a time that we retire that conversion.

rengolin commented on lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:
```diff
@@ -287,6 +358,175 @@ FailureOr<BinaryFlags> getBinaryFlags(Type operandType, Type outputType,
+    OpBuilder::InsertionGuard guard(rewriter);
+    // Insert before module terminator.
+    rewriter.setInsertionPoint(module.getBody(),
+                               std::prev(module.getBody()->end()));
```
Code repetition, please adapt `BuilderUtils::createFunction`.

KavithaTipturMadhu replied:
I didn't realize this immediately, but we miss out on licm for dispatch calls if we adapt BuilderUtils::createFunction in place of this function. What do you suggest we do here?

rengolin commented on lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:
```diff
@@ -287,6 +358,175 @@ FailureOr<BinaryFlags> getBinaryFlags(Type operandType, Type outputType,
+  }
+
+  OpBuilder::InsertionGuard guard(rewriter);
+  rewriter.setInsertionPointAfter(dispatchOperands.back().getDefiningOp());
```
This doesn't seem necessary. If you're building this function now, then all the producers have already been lowered into IR and the insertion point is at the right place?

KavithaTipturMadhu replied:
Function creation is right before, so that moves the insertion point. Thats why I think I set the insertion point explicitly, nevertheless I'll verify if I can remove this line safely.

KavithaTipturMadhu replied:
Removed the insertion point, as this wasnt necessary as rightly pointed out by Renato.

rengolin commented on lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:
```diff
@@ -287,6 +358,175 @@ FailureOr<BinaryFlags> getBinaryFlags(Type operandType, Type outputType,
+  FlatSymbolRefAttr fnName =
+      SymbolRefAttr::get(rewriter.getContext(), invokeName);
+
+  if (!module.lookupSymbol(fnName)) {
```
Code repetition, please adapt `BuilderUtils::createFunction`.

KavithaTipturMadhu replied:
Done.

rengolin commented on lib/TPP/Conversion/ConvertVectorToXsmm/ConvertTranspose.cpp:
```diff
@@ -0,0 +1,253 @@
+  IntegerType integer64 = IntegerType::get(rewriter.getContext(), 64);
+  dispatchOperands.push_back(rewriter.create<arith::ConstantOp>(
+      loc, integer64, cast<TypedAttr>(kind)));
+  dispatchOperandTypes.push_back(integer64);
```
Can't you really get the type from the value list?

KavithaTipturMadhu replied:
I'll change it to get the type from the value list.

KavithaTipturMadhu replied:
Done.

rengolin commented on lib/TPP/Conversion/ConvertVectorToXsmm/ConvertTranspose.cpp:
```diff
@@ -0,0 +1,253 @@
+  xsmm::UnaryKindAttr kind =
+      xsmm::UnaryKindAttr::get(rewriter.getContext(), unaryKind);
+  IntegerType integer64 = IntegerType::get(rewriter.getContext(), 64);
+  dispatchOperands.push_back(rewriter.create<arith::ConstantOp>(
```
`BuilderUtils::getConstInt`

rengolin commented on lib/TPP/GPU/LinalgToXeGPU.cpp:
```diff
@@ -196,7 +196,7 @@ static std::optional<Value> lowerGenericOp(linalg::GenericOp genericOp,
 
     auto zeroVec =
-        rewriter.create<vector::BroadcastOp>(loc, resType, zeroConst);
+        rewriter.create<mlir::vector::BroadcastOp>(loc, resType, zeroConst);
```
If this change is necessary now (it wasn't before), then you changed something in the namespace resolution. Make sure you remove the change and these lines can continue as they were.

Always revisit your assumptions when a file that should remain untouched needs to be touched. Often you're doing something unnecessary elsewhere.

KavithaTipturMadhu replied:
My bad, that was a stray change, I'll get rid of it.

KavithaTipturMadhu replied:
I have reverted the changes to both LinalgToXeGPU and GPUVectorize, theyre no longer needed as I've changed the dialect dependence to mlir::vector::VectorDialect instead of vector::VectorDialect.

rengolin commented on lib/TPP/Transforms/Utils/VNNIUtils.cpp:
```diff
@@ -69,6 +69,17 @@ FailureOr<AffineDimExpr> isInVnniLayout(linalg::GenericOp linalgOp,
 }
 
+bool isInVnniLayout(VnniOperandRank expectedRank, VectorType vector) {
+  return isInVnniLayout((int64_t)expectedRank, vector);
```
this wrapper seems unnecessary, this C-style cast is not appropriate.

KavithaTipturMadhu replied:
Agree, will get rid of it.

rengolin commented on test/Conversion/VectorToXsmm/vector-to-transpose.mlir:
```diff
@@ -0,0 +1,129 @@
+// CHECK:   call @xsmm_unary_invoke(%[[c2_i64]], %[[dispatch]], %[[inttoptr]], %[[offset]], %[[inttoptr2]], %[[offset_1]])
+
+// -----
+func.func @not_vnni_packing_1(%arg0: memref<32x32xf32, strided<[512, 1], offset: ?>>, %arg1: memref<16x32x2xf32, strided<[64, 2, 1], offset: ?>>) {
```
took me a while to realize this was `f32`. Please update the function name with `not_vnni_in_f32` or something.

KavithaTipturMadhu replied:
Done.

rengolin commented on test/Integration/tpp-transpose.mlir:
```diff
@@ -0,0 +1,15 @@
+
+module {
+  func.func @entry(%arg0: tensor<3x5xf32>, %arg1: tensor<5x3xf32>)->tensor<5x3xf32> {
+    %out = linalg.transpose ins(%arg0 : tensor<3x5xf32>) outs(%arg1 : tensor<5x3xf32>) permutation = [1, 0]
```
From my interpretation, your pattern matching above needs vector read/write to match the pattern. Will this really be converted into XSMM calls?

Can you add another RUN like that prints the MLIR=late and CHECK for xsmm calls?

KavithaTipturMadhu replied:
Will add print-mid/print-late

KavithaTipturMadhu replied:
Done.

rengolin commented on tools/tpp-opt/tpp-opt.cpp:
```diff
@@ -30,6 +30,7 @@
 #include "TPP/Dialect/Xsmm/XsmmDialect.h"
 #include "TPP/PassBundles.h"
 #include "TPP/Passes.h"
+#include "TPP/Passes.h.inc"
```
Nope. If you need this, something is really wrong elsewhere.

KavithaTipturMadhu replied:
This isn't necessary, I forgot to remove this.

KavithaTipturMadhu replied:
Done.

rengolin commented on test/Integration/tpp-transpose-bf16.mlir:
```diff
@@ -0,0 +1,525 @@
+// RUN: tpp-run --vector-To-XSMM %s -e entry -entry-point-result=void -print --seed 123 2>&1 | FileCheck %s
+// RUN: tpp-run --linalg-to-loops %s -e entry -entry-point-result=void -print --seed 123 2>&1 | FileCheck %s
```
Same comment as the other integration test: please add a RUN line to make sure we continue to match vector to XSMM.

KavithaTipturMadhu replied:
Done.

rengolin commented on test/Integration/tpp-transpose-bf16.mlir:
```diff
@@ -0,0 +1,525 @@
+// CHECK:	( 0.207031, 0 )
+// CHECK:	( 0, 0.0991211 )
+// CHECK:	( 0.0649414, 0.242188 )
+// CHECK:	( 0.154297, 0.171875 )
```
you don't need to check against ALL values, just the first dozen or so.

KavithaTipturMadhu replied:
I didnt see this comment before, we might end up passing incorrect size arguments to xsmm call which might go unnoticed if we dont check ALL values. Am i missing something?

KavithaTipturMadhu replied:
@rengolin what do you think? Am i missing something here?

KavithaTipturMadhu replied:
as discussed, I have reduced the problem size so that the results are smaller and can be fully compared.

KavithaTipturMadhu commented on test/Conversion/VectorToXsmm/vector-to-transpose.mlir:
```diff
@@ -0,0 +1,129 @@
+// RUN: tpp-opt --vector-to-xsmm %s --split-input-file | FileCheck %s
+
+func.func @transpose_op_0(%arg0: memref<3x5xf32>, %arg1: memref<5x3xf32>) {
```
Done.

KavithaTipturMadhu commented:
I have mostly addressed all the review comments as per my understanding. I havent marked the conversations as closed yet since they may have opens in your opinion @rengolin Please review it wn you have the time. The CI is green.

rengolin commented on lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:
```diff
@@ -8,18 +8,63 @@
+#define DEBUG_TYPE "xsmm-utils"
+
+using namespace mlir::linalg;
+using namespace mlir::structured_match;
```
Do you really need this?

rengolin commented on include/TPP/Dialect/Xsmm/XsmmUtils.h:
```diff
@@ -11,7 +11,9 @@
 #include "TPP/Dialect/Xsmm/XsmmEnum.h"
 #include "TPP/Dialect/Xsmm/XsmmOps.h"
-#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "TPP/IR/StructuredOpMatcher.h"
```
Do not add more headers to headers if you're not using it in the actual header. Add it to the CPP file that uses it.

rengolin commented on include/TPP/Passes.td:
```diff
@@ -54,7 +54,7 @@ def VectorizationPass : Pass<"vectorization-pass",
   }];
 
-  let dependentDialects = [ "memref::MemRefDialect", "linalg::LinalgDialect", "vector::VectorDialect" ];
+  let dependentDialects = [ "memref::MemRefDialect", "linalg::LinalgDialect", "mlir::vector::VectorDialect" ];
```
It's still unclear to me why you need this now since we didn't before. There is no clash with `std::vector`.

rengolin commented on lib/TPP/Transforms/Utils/BuilderUtils.cpp:
```diff
@@ -35,13 +35,15 @@ arith::ConstantOp getConstant(OpBuilder &builder, Type type, ValueT value) {
 
 func::FuncOp createFunction(OpBuilder &builder, ModuleOp module, StringRef name,
-                            TypeRange args, TypeRange ret) {
+                            TypeRange args, TypeRange ret, bool createBody) {
```
IIRC, your removed code had private visibility, while this function gives public visibility. It should not matter for our purposes, just pointing out because you manually set it. Making sure you're not missing anything.

KavithaTipturMadhu replied:
I didnt want to change the API.I can either add an argument, similarly as the createBody flag, or just set the function as private after creating the prototype. [Here's why](https://mlir.llvm.org/getting_started/Faq/#whats-with-op-symbol-declaration-cannot-have-public-visibility). What do you recommend?

rengolin commented on lib/TPP/Conversion/ConvertVectorToXsmm/ConvertVectorToXsmmPDL.pdll:
```diff
@@ -0,0 +1,22 @@
+
+Constraint ValidateTranspose(op:Op<vector.transpose>, input0:Op<vector.transfer_read>, output:Op<vector.transfer_write>, outputType:TypeRange);
+
+Pattern ConvertTransposeiPattern{
```
typo

KavithaTipturMadhu replied:
Will fix.

rengolin commented on lib/TPP/Conversion/ConvertVectorToXsmm/ConvertVectorToXsmmPDL.pdll:
```diff
@@ -0,0 +1,22 @@
+     rewrite transpose with{
+        let replacement = ConvertTranspose(transpose, input0, output0, transposeOutput0);
+        replace transpose with (replacement.dispatch, replacement.invoke);
+        erase output0;
```
shouldn't it erase `input0` too?

KavithaTipturMadhu replied:
input0 will get dced.

rengolin commented on lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:
```diff
@@ -287,6 +354,76 @@ FailureOr<BinaryFlags> getBinaryFlags(Type operandType, Type outputType,
+  func::CallOp call;
+  //  if (insertAfter) {
+  //  rewriter.setInsertionPointAfter(insertAfter);
+  if (insertBefore) {
```
Avoid the pattern of long if/else blocks. Use the if/else to build the argument (`operands`) then emit a single `call = ` line after the block.

rengolin commented on lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:
```diff
@@ -287,6 +354,76 @@ FailureOr<BinaryFlags> getBinaryFlags(Type operandType, Type outputType,
+                                      IntegerAttr dataTypeAttr,
+                                      Operation *parentOp) {
+  SmallVector<Value> res;
+  builder.setInsertionPoint(parentOp);
```
Please do _not_ use `setInsertionPoint` without an insertion guard.

rengolin commented on lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:
```diff
@@ -287,6 +354,76 @@ FailureOr<BinaryFlags> getBinaryFlags(Type operandType, Type outputType,
+  //  rewriter.setInsertionPointAfter(insertAfter);
+  if (insertBefore) {
+    rewriter.setInsertionPoint(insertBefore);
+    auto finalOperands = xsmm::utils::getOperands(
```
This seems an arbitrary decision?

Why is it that, when `insertBefore` is non null, you extract the operands, but when it's null, you don't?

Is this just the difference between `dispatch` and `invoke`? If so, then this should be explicit in the code and not veiled behind the existence of an unrelated (to this code at least) argument.

KavithaTipturMadhu replied:
I'll add a flag for type, in order to distinguish between dispatch and invoke types.

rengolin commented on tools/tpp-run/tpp-run.cpp:
```diff
@@ -147,6 +147,10 @@ llvm::cl::opt<bool>
                llvm::cl::desc("Kernel buffers are allocated on GPU"),
                llvm::cl::init(true));
 
+llvm::cl::opt<bool> defVectorToXSMM("vector-to-xsmm",
```
Please do not put this here. This is the wrong place. None of the other options are here because this is not a tool option, but a pipeline option. All of the other pipeline options are in the pipeline code. You have already mentioned `lhsTile` and `rhsTile`, and they are not here either.

KavithaTipturMadhu replied:
lhsTile and rhsTile are explicitly added as pipeline options and are not bundles, as you pointed out yourself. gpubackend is a bundle and is added this way, and the way to add bundles is this. The only other way to add options to DefaultPipeline is like lhsTile, which is with a different name, which does not match vector-to-xsmm.

rengolin commented on lib/TPP/Conversion/ConvertVectorToXsmm/ConvertVectorToXsmm.cpp:
```diff
@@ -0,0 +1,235 @@
+  auto functionOp = transposeOp->getParentOfType<func::FuncOp>();
+  SmallVector<Value, 10> dispatchOperands;
+  OpBuilder::InsertionGuard guard(rewriter);
+  rewriter.setInsertionPoint(&*functionOp.getBody().op_begin());
```
This change of insertion point doesn't need to be this early. Move it down with the use (`buildXsmmCall`) so that it's clear why you're changing it.

KavithaTipturMadhu replied:
We'll then have to change the insertion point conditionally, as we're using buildXsmmCall for both invokes and dispatches, which I feel is confusing. What do you think?

rengolin commented on lib/TPP/Conversion/ConvertVectorToXsmm/ConvertVectorToXsmm.cpp:
```diff
@@ -0,0 +1,235 @@
+      xsmm::UnaryFlagsAttr::get(rewriter.getContext(), xsmm::UnaryFlags::NONE));
+  dispatchOperands.push_back(getConstInt(rewriter, unaryFlags.getInt(), 64));
+
+  auto dispatched = xsmm::utils::buildXsmmCall(
```
As discussed in the definition, `buildXsmmCall` is ambiguous. You need to make it clear which of them you're changing it (dispatch, invoke).

If you have two calls (or a type argument), then you don't even need to change the insertion point in this function. You can set it for _dispatch_ and the insertion guard will make it sure that when you build the `invoke`, it will be back in the right place.

rengolin commented on lib/TPP/Conversion/ConvertVectorToXsmm/ConvertVectorToXsmm.cpp:
```diff
@@ -0,0 +1,235 @@
+        stridesOnOutput->front() /
+        *vnni::utils::getVnniBlockingFactor(expandShapeOp.getSrcType());
+    auto functionOp = transposeOp->getParentOfType<func::FuncOp>();
+    OpBuilder::InsertionGuard guard(rewriter);
```
You're again setting the insertion point without reason and guards. Please do _not_ do that. This is very dangerous.

KavithaTipturMadhu replied:
I was trying to ensure that constants are created at the beginning of the function body. I didn't realize that the default insertion point would be the beginning of the function body. Should we not set the guard and insertion point explicitly to ensure the insertion point is set correctly or should we rely on default insertion points?

KavithaTipturMadhu replied:
Will delete.

rengolin commented on lib/TPP/Conversion/ConvertVectorToXsmm/ConvertVectorToXsmm.cpp:
```diff
@@ -0,0 +1,235 @@
+        dyn_cast<VectorType>(input->getResult(0).getType()),
+        dyn_cast<VectorType>(output->getOperand(0).getType()),
+        xsmm::UnaryFlags::NONE);
+    auto temp = unaryInfo->n;
```
`std::swap`?

rengolin commented on lib/TPP/Conversion/ConvertVectorToXsmm/ConvertVectorToXsmm.cpp:
```diff
@@ -0,0 +1,235 @@
+    unaryInfo->ldi = unaryInfo->ldo;
+    unaryInfo->ldo = temp;
+
+    return getTransposeOperation(rewriter, *unaryInfo, outputType,
```
This is still duplicated code: `getTransposeOperation` is called with basically the same arguments on both sides of the branch.

As discussed before, use the if/else block to define the arguments then call the function only once.

rengolin commented on lib/TPP/Conversion/ConvertVectorToXsmm/ConvertVectorToXsmm.cpp:
```diff
@@ -0,0 +1,235 @@
+        dyn_cast<memref::ExpandShapeOp>(source.getDefiningOp());
+    source = expandShapeOp.getSrc();
+    xsmm::UnaryInfo unaryInfo;
+    unaryInfo.m = expandShapeOp.getSrcType().getShape()[0];
```
Why isn't this logic in `getVectorUnaryInfo`?

KavithaTipturMadhu replied:
I thought since the code was specific to vnni and wasnt gonna be used elsewhere, there was no need to move it to getVectorUnaryInfo. I'll move it there.

rengolin commented on lib/TPP/Conversion/ConvertVectorToXsmm/ConvertVectorToXsmm.cpp:
```diff
@@ -0,0 +1,235 @@
+  return success();
+}
+
+void registerNativeRewrite(RewritePatternSet &patterns) {
```
```suggestion
static void registerNativeRewrite(RewritePatternSet &patterns) {
```

rengolin commented:
> I have mostly addressed all the review comments as per my understanding. I havent marked the conversations as closed yet since they may have opens in your opinion @rengolin Please review it wn you have the time. The CI is green.

Did you forget to push? I'm not seeing your changes.

KavithaTipturMadhu commented:
> > > > I have mostly addressed all the review comments as per my understanding. I havent marked the conversations as closed yet since they may have opens in your opinion @rengolin Please review it wn you have the time. The CI is green.
> > > 
> > > 
> > > Did you forget to push? I'm not seeing your changes.
> > 
> > 
> > This was my reply yesterday, I'm still working on a few more comments, gimme an hour or so, will push my changes.
> 
> I see. Then please avoid requesting reviews before pushing everything you have. I spent some time today working around Github thinking it was broken that it wasn't showing any changes. Try to time your reply _after_ you make the final push.

I commented this yesterday and thought you posted more comments afterwards, since your comments were on the latest changes. I was referring to the changes I was going to make based on those comments before I made the next push.

KavithaTipturMadhu commented:
@rengolin I think I have addressed all the comments barring the one which said I shouldn't add vector-to-xsmm flag to tpp-run.cpp. Please do take a look when you have the time. I apologize for the delay. Also, if I have missed something, my sincere apologies.

rengolin commented on lib/TPP/Conversion/ConvertVectorToXsmm/ConvertVectorToXsmm.cpp:
```diff
@@ -0,0 +1,220 @@
+}
+
+static LogicalResult
+validateTransposeOpImpl(PatternRewriter &rewriter, Operation *transposeOp,
```
Why `Impl` in the name? This is a local function that just validates something. `Impl` is a pattern in LLVM/MLIR (and general C++ projects) that imply it's the _hidden_ part of a public API that is not accessible to the users of that API.

Here, there's no API.

KavithaTipturMadhu commented on lib/TPP/Conversion/ConvertVectorToXsmm/ConvertVectorToXsmm.cpp:
```diff
@@ -0,0 +1,220 @@
+    unaryInfo.ldo = unaryInfo.ldo * vnniFactor;
+    unaryInfo.ldi = unaryInfo.ldi / vnniFactor;
+    std::swap(unaryInfo.ldi, unaryInfo.ldo);
+  } else
+    std::swap(unaryInfo.m, unaryInfo.n);
```
I thought the llvm coding standard said that we should not use a brace for single statement bodies, atleast thats what I remember Lorenzo telling me: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

adam-smnk replied:
That's true if all cases are simple one liners. When any requires braces (be it more code, comments etc.), all the other cases should have braces too to keep whole code block uniform.

KavithaTipturMadhu commented on lib/TPP/PassBundles/VectorToXSMM.cpp:
```diff
@@ -6,10 +6,19 @@
 
+#include "TPP/Conversion/ConvertVectorToXsmm/ConvertVectorToXsmm.h"
 #include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
```
I have trimmed the list to a minimum.

rengolin replied:
I'm still surprised with the number. @adam-smnk I just copy this file from the other examples. Now I noticed we include `"TPP/PassBundles.h.inc"`, which sounds wrong (why not `"TPP/PassBundles.cpp.inc"`?).

Not for this PR, but we need to have a harder look on these bundles again soon.

rolfmorel commented on include/TPP/PassBundles.td:
```diff
@@ -10,6 +10,7 @@
 #define TPP_DIALECT_TPP_PASSBUNDLES
 
 include "mlir/Pass/PassBase.td"
+include "mlir/Dialect/PDL/IR/PDLDialect.td"
```
Is this needed? I expect you don't need any symbols from this file in `PassBundles.td`.

KavithaTipturMadhu replied:
You're right, removed.

KavithaTipturMadhu commented:
@rengolin I have addressed all comments, I haven't updated the integration tests the way we discussed but just kept the xsmm calls as check lines because we have a new flag along with linalg-to-vector flag which forces lowering to vector. I can add check lines to ensure that we lower to vector in the first place. What do you think?

rengolin commented on lib/TPP/DefaultTppPasses.cpp:
```diff
@@ -90,8 +90,10 @@ struct DefaultTppPasses
     }
-    if (vectorToXSMM)
-      skipOperations.clear();
+    if (vectorToXSMM) {
```
Careful. `linalg-to-vector` and `vector-to-xsmm` could both be set, and this would result in a vector `{ "all", "transpose", "vnni" }`.

KavithaTipturMadhu replied:
Fixed.

rengolin commented on test/Integration/transpose-fp32.mlir:
```diff
@@ -0,0 +1,21 @@
+}
+
+
+// XSMM: call @xsmm_unary_dispatch
```
As discussed, let's leave this test as is, but create a new one that dumps the IR between passes and makes sure we get vector and xsmm in that order.

rengolin commented on lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:
```diff
@@ -8,18 +8,60 @@
+#include "TPP/Transforms/Utils/VNNIUtils.h"
 #include "TPP/Transforms/Utils/ValueUtils.h"
 #include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
```
I'm surprised by the need of the LLVM dialect header here.

KavithaTipturMadhu replied:
It's for memref expansion into llvm ops.

rengolin reviewed (commented):
Looking much better, thanks! Added a few more comments but nothing major. Just need to make sure we have the test that shows your pass is actually running and looks good.

rengolin commented on lib/TPP/Conversion/ConvertVectorToXsmm/ConvertVectorToXsmm.cpp:
```diff
@@ -0,0 +1,218 @@
+                                         Operation *transposeOp,
+                                         Operation *input, Operation *output,
+                                         Type outputType) {
+  LLVM_DEBUG(llvm::dbgs() << "validateTransposeOpImpl\n");
```
Forgot one last `Impl` from the comment.

rengolin commented on include/TPP/PassBundles.td:
```diff
@@ -41,6 +41,9 @@ def DefaultTppPasses : Pass<"default-tpp-passes", "ModuleOp"> {
     Option<"vectorToXSMM", "vector-to-xsmm",
            "bool", /*default=*/"false",
            "Lower vector patterns to XSMM calls.">,
+    Option<"printVector", "print-vector-ir",
```
This is not needed anymore

rengolin reviewed (commented):
Only one left-over and looks good, thanks!

rengolin reviewed (approved):
LGTM, thank you!
